modules/auxiliary/admin/citrix/citrix_netscaler_config_decrypt.rb

```diff
@@ -0,0 +1,259 @@
+        'DisclosureDate' => '2022-05-19',
+        'License' => MSF_LICENSE,
+        'References' => [
+          ['URL', 'https://dozer.nz/posts/citrix-decrypt/'],
+          ['URL', 'https://www.ferroquesystems.com/resource/citrix-adc-security-kek-files/']
+        ],
+        'Actions' => [
+          [
+            'Dump',
+            {
+              'Description' => 'Dump secrets from NetScaler configuration'
+            }
+          ]
+        ],
+        'DefaultAction' => 'Dump',
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+
+    register_options([
+      OptPath.new('NS_CONF', [ true, 'Path to a NetScaler configuration file (ns.conf)' ]),
+      OptPath.new('NS_KEK_F1', [ false, 'Path to NetScaler KEK fragment file F1.key' ]),
+      OptPath.new('NS_KEK_F2', [ false, 'Path to NetScaler KEK fragment file F2.key' ]),
+      OptString.new('NS_IP', [ false, '(Optional) IPv4 address to attach to loot' ])
+    ])
+  end
+
+  def loot_host
+    datastore['NS_IP'] || '127.0.0.1'
+  end
+
+  def ns_conf
+    datastore['NS_CONF']
+  end
+
+  def ns_kek_f1
+    datastore['NS_KEK_F1']
+  end
+
+  def ns_kek_f2
+    datastore['NS_KEK_F2']
+  end
+
+  # ns.conf elements that contain potential secrets, update as needed
+  # k = parameter that has the secret (-key, -password, [...])
+  # v = start of config line that potentially has a secret
+  def ns_secret
+    {
+      'key' => ['add ssl certKey'],
+      'keyValue' => ['set ns encryptionParams'],
+      'radKey' => ['add authentication radiusAction'],
+      'ldapBindDnPassword' => ['add authentication ldapAction'],
+      'password' => ['set ns rpcNode', 'add lb monitor', 'add aaa user'],
+      'passPhrase' => ['add authentication dfaAction']
+    }
+  end
+
+  # Statically defined in libnscli90.so, modern appliances keep these in /nsconfig/.skf
+  def ns90_rc4key
+    '2286da6ca015bcd9b7259753c2a5fbc2'.scan(/../).map(&:hex).pack('C*')
+  end
+
+  def ns90_aeskey
+    '351cbe38f041320f22d990ad8365889c7de2fcccae5a1a8707e21e4adccd4ad9'.scan(/../).map(&:hex).pack('C*')
+  end
+
+  def run
+    if ns_kek_f1 && ns_kek_f2
+      print_status('Building NetScaler KEK from key fragments ...')
+      unless build_ns_kek
+        fail_with(Msf::Exploit::Failure::NoTarget, 'Unable to build NetScaler KEK from key fragments')
+      end
+    end
+    parse_ns_config
+  end
+
+  def build_ns_kek
+    unless File.size(ns_kek_f1) == 256 && File.size(ns_kek_f2) == 256
+      print_error('KEK files must be 256 bytes in size')
+      return false
+    end
+    f1_hex = File.read(ns_kek_f1)
+    f2_hex = File.read(ns_kek_f2)
+    f1_hex&.encode('UTF-8', 'binary')
+    f2_hex&.encode('UTF-8', 'binary')
+    unless f1_hex.match?(/^[0-9a-f]+$/i) && f2_hex.match?(/^[0-9a-f]+$/i)
+      print_error('NS KEK files must be hexadecimal format')
+      return false
+    end
```

Comment:
When then files aren't able to be processed, the module should probably `raise Msf::OptionValidateError.new(['OPTION_NAME'])`.

Reply:
I can dance to this.